mne/preprocessing/hfc.py

```diff
@@ -0,0 +1,135 @@
+from .maxwell import _prep_mf_coils, _sss_basis
+from ..forward import _prep_meg_channels
+from ..io.pick import pick_types
+from ..io.proj import Projection
+from ..utils import verbose
+
+
+@verbose
+def compute_proj_hfc(info, order=1, picks=None, exclude_bads=True,
+                     accuracy='accurate', verbose=None):
+    """Generate projectors to perform homogeneous/harmonic correction to data.
+
+    Remove evironmental fields from magentometer data by assuming it is
+    explained as a homogeneous :footcite:`TierneyEtAl2021` or harmonic field
+    :footcite:`TierneyEtAl2022`. Useful for arrays of OPMs.
+
+    Parameters
+    ----------
+    info : instance of Info
+        Info from an instance of Raw (preferable).
+    order : int
+        The order of the spherical harmonic basis set to use. Set to 1 to use
+        only the homogeneous field component (default), 2 to add gradients, 3
+        to add quadrature terms etc.
+    picks : int | array-like
+        List of channels indices to compute projeciton on. Default behaviour
+        is to select meg (non-reference) with known positions.
+    exclude_bads : bool
+        Do not include bad channels in the projection, (default: True).
+    accuracy : str
+        Can be ``"point"``, ``"normal"`` or ``"accurate"`` (default), defines
+        which level of coil definition accuracy is used to generate model.
+    %(verbose)s
+
+    Returns
+    -------
+    projs: list
+        List of projection vectors.
+
+    Notes
+    -----
+    To apply the projectors to a dataset, use
+    ``[DATA].add_proj(projs).apply_proj()``.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    if picks is None:
+        if exclude_bads is True:
+            picks = pick_types(info, meg=True)
+        else:
+            picks = pick_types(info, meg=True, exclude=None)
+    picks = _filter_channels_with_positions(info, picks)
+    basis, channels = _generate_basis_set(info, picks, order, accuracy)
+    labels = _label_basis(order)
+    projs = []
+    for ii in range(len(labels)):
+        data = basis[:, ii]
+        proj_data = dict(col_names=channels, row_names=None,
+                         data=data[np.newaxis, :], ncol=len(channels), nrow=1)
+        proj = Projection(active=False, data=proj_data, desc=labels[ii])
+        projs.append(proj)
+
+    return projs
+
+
+def _filter_channels_with_positions(info, indsin):
+    """Keep indices of channels with position information."""
+    ch_inds = list()
+    for ii in indsin:
+        if not (any(np.isnan(info['chs'][ii]['loc']))):
+            ch_inds.append(ii)
+    return ch_inds
+
+
+def _generate_basis_set(info, picks=None, order=1,
+                        accuracy='accurate', origin=(0, 0, 0)):
+    """Generate the basis set used for HFC."""
+    exp = dict(origin=(0, 0, 0), int_order=0, ext_order=order)
+    coils = _prep_mf_coils(info, accuracy=accuracy)
+    mf_names = _get_mf_names(info, accuracy=accuracy)
```

Comment:
Why are `mf_names` necessary? It seems like all channels should be used...

Reply:
It shouldn't be necessary if all (non-ref) channels are used, but if a user sub-selects some channels, _prep_mf_coils doesnt return any channels names so we can't trivially map between the picks and the coils. I did think about making _prep_mf_coils to return names too, but I think that will break most of the maxfilter code. So mf_names is a bit of a sticking plaster.

Reply:
> . I did think about making _prep_mf_coils to return names too, but I think that will break most

How about adding a new return_names=False kwarg that you set to True for your new from?